File: languages/go/BinarySearch.go
// Binary Search: Divides sorted array in half repeatedly to find target
// Time: O(log n), Space: O(1)

package main

import (
    "fmt"
    "sort"
)

func binarySearch(arr []int, target int) int {
    left, right := 0, len(arr)-1

    for left <= right {
        mid := left + (right-left)/2
        if arr[mid] == target {
            return mid
        } else if arr[mid] < target {
            left = mid + 1
        } else {
            right = mid - 1
        }
    }

    return -1
}

func main() {
    data := []int{164, 29, 7, 190, 71, 63, 58, 36, 189, 27, 174, 140, 23, 152, 109, 9, 8, 24, 56, 60, 130, 155, 198, 144, 51, 167, 192, 108, 57, 115, 151, 72, 2, 41, 186, 88, 169, 40, 182, 87, 191, 183, 98, 25, 92, 89, 68, 12, 118, 138}
    sort.Ints(data)
    target := 88
    result := binarySearch(data, target)
    fmt.Printf("Target %d found at index: %d\n", target, result)
}

File: languages/java/BinarySearch.java
// Binary Search: Divides sorted array in half repeatedly to find target
// Time: O(log n), Space: O(1)

import java.io.IOException;
import java.util.Arrays;

public class BinarySearch {
    public static int binarySearch(int[] arr, int target) {
        int left = 0, right = arr.length - 1;
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (arr[mid] == target) {
                return mid;
            } else if (arr[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        
        return -1;
    }

    public static void main(String[] args) throws IOException {
        int[] data = DatasetReader.readDataset("../large_dataset.txt");
        Arrays.sort(data);
        int target = 88;
        int result = binarySearch(data, target);
        System.out.println("Target " + target + " found at index: " + result);
    }
}